### PR TITLE
[train] Simplify `ray.train.xgboost/lightgbm` (3/n): Re-implement `LightGBMTrainer` as a lightweight `DataParallelTrainer`

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR re-implements `LightGBMTrainer` as a `DataParallelTrainer` that does not use `lightgbm_ray` under the hood, in an effort to unify the trainer implementations and remove that external dependency.

This PR is the `lightgbm` counterpart to https://github.com/ray-project/ray/pull/42767. See that PR's description for details.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/_internal/data_config.py
Comment (matthewdeng) [inline python/ray/train/_internal/data_config.py:1]: Revert these changes in favor of materialize?

File: python/ray/train/lightgbm/config.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    Ray Train will set up the necessary coordinator processes and environment
+    variables for your workers to communicate with each other.
+
+    See the LightGBM docs for more information on the "network parameters"
+    that Ray Train sets up for you:
+    https://lightgbm.readthedocs.io/en/latest/Parameters.html#network-parameters
+    """
+
+    NETWORK_PARAMS_KEY = "LIGHTGBM_NETWORK_PARAMS"
```
Comment (matthewdeng) [inline python/ray/train/lightgbm/config.py:24]: nit: Any reason this is defined in the Config rather than globally somewhere?

File: python/ray/train/lightgbm/lightgbm_trainer.py [hunk 2]
```diff
@@ -1,15 +1,223 @@
...
+    train_X, train_y = train_df.drop(label_column, axis=1), train_df[label_column]
+    train_set = lightgbm.Dataset(train_X, label=train_y)
+    valid_sets = []
+    for eval_df in eval_dfs.values():
+        eval_X, eval_y = eval_df.drop(label_column, axis=1), eval_df[label_column]
+        valid_sets.append(lightgbm.Dataset(eval_X, label=eval_y))
+    valid_names = list(eval_dfs.keys())
+
+    lightgbm.train(
+        params=config,
```
Comment (matthewdeng) [inline python/ray/train/lightgbm/lightgbm_trainer.py:63]: Needs to include `get_network_params()`? 😄

**Review by matthewdeng — COMMENTED** [review]
Neato!

File: python/ray/train/lightgbm/config.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/lightgbm/config.py:24]: No reason in particular 😄  do you prefer just as a constant in this file?

File: python/ray/train/lightgbm/lightgbm_trainer.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/train/lightgbm/lightgbm_trainer.py:63]: yes good catch!

Comment (matthewdeng) [reply python/ray/train/lightgbm/lightgbm_trainer.py:63]: Make sure this actually runs distributed training 😏

File: python/ray/train/lightgbm/config.py (hunk 1 above)
Comment (matthewdeng) [reply python/ray/train/lightgbm/config.py:24]: I guess it depends how much this ends up being exposed to users. I see there's an assertion that mentions it, but the dependency between the Config <> Backend <> constant might be confusing for users.

File: python/ray/train/lightgbm/lightgbm_trainer.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/train/lightgbm/lightgbm_trainer.py:67]: Yeah, I will need to add back a test suite like this one to make sure distributed training is actually happening properly.

https://github.com/ray-project/xgboost_ray/blob/5a840af05d487171883dadbfdd37b138b607bed8/xgboost_ray/tests/test_end_to_end.py#L56

For now I have confirmed it by running some examples.

File: python/ray/train/lightgbm/__init__.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
 from ray.train.lightgbm._lightgbm_utils import RayTrainReportCallback
+from ray.train.lightgbm.config import LightGBMConfig
 from ray.train.lightgbm.lightgbm_checkpoint import LightGBMCheckpoint
 from ray.train.lightgbm.lightgbm_predictor import LightGBMPredictor
 from ray.train.lightgbm.lightgbm_trainer import LightGBMTrainer
 
 __all__ = [
     "RayTrainReportCallback",
     "LightGBMCheckpoint",
+    "LightGBMConfig",
```
Comment (woshiyyya) [inline python/ray/train/lightgbm/__init__.py:10]: nit: Are we exposing this `LightGBMConfig` as a public API?  Seems no special logics in this config, but it's still an argument in the `LightGBMTrainer`.

**Review by woshiyyya — APPROVED** [review]
LGTM : )

File: python/ray/train/lightgbm/__init__.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/train/lightgbm/__init__.py:10]: Hmm, yeah there's not really a reason for a user to know about `LightGBMConfig`.

I was trying to keep the API consistent for `LightGBMTrainer` with all the other `DataParallelTrainer`s that take in a backend config. So, `LightGBMConfig` would need to be public for it to be a constructor argument of `v2.LightGBMTrainer`.

Let me remove it from the public `ray.train.lightgbm` namespace for now while `v2.LightGBMTrainer` is still private.

**Review by matthewdeng — APPROVED** [review]
<div><br/></div>
